### PR TITLE
Ensure the initial RAM disk address is suitable

### LINKED ISSUE #3628: Add checks for the suggested initial RAM disk address to make sure it is valid and won't cause problems
Currently we just trust the address suggested by the hypervisor to load the initial RAM disk into memory. Since we only add an identity mapping for the first 1GiB of RAM, we should check that it falls in this range and also will not overwrite the kernel or the important structures below 2MiB.

### PR DESCRIPTION
**Pull request by conradgrobler**
Checks whether the suggested initial RAM disk address is suitable, and if not tries to find a more suitable one.

Fixes #3628

### CODE REVIEW

File: stage0/src/initramfs.rs [hunk 1]
```diff
@@ -16,38 +16,135 @@
...
-    let buf = unsafe { slice::from_raw_parts_mut::<u8>(address.as_mut_ptr(), size as usize) };
+    // We use an identity mapping for memory.
+    let address = VirtAddr::new(initrd_address.as_u64());
+
+    // Safety: We already checked that the slice falls within the mapped virtual memory, is backed
+    // by physical memory, does not overlap with the kernel, and does not overlap with the first
+    // 2MiB where our important data structures live. It can also not overlap with the firmware ROM
+    // image, since that is above the 1GiB upper limit we support. We only write to the slice and
+    // don't assume anything about its layout or alignment.
+    let buf = unsafe { slice::from_raw_parts_mut::<u8>(address.as_mut_ptr(), size) };
```
Comment (andrisaar) [inline stage0/src/initramfs.rs:91]: We should probably mark the memory as reserved in the E820 map as well, for both (a) safety that we won't stomp on the data structure later, and (b) ensuring that the kernel itself doesn't accidentally overwrite the memory.

(Or, check what SeaBIOS and/or the linuxboot option ROMs are doing -- do they mark the memory as reserved?)

File: stage0/src/initramfs.rs [hunk 2]
```diff
@@ -16,38 +16,135 @@
...
         .expect("couldn't read initial RAM disk address");
 
-    // We use an identity mapping for memory.
-    let address = VirtAddr::new(address.as_u64());
-
     log::debug!("Initial RAM disk size {}", size);
-    log::debug!("Initial RAM disk address {:#018x}", address.as_u64());
+    log::debug!("Initial RAM disk address {:#018x}", initrd_address.as_u64());
+
+    let kernel_address = fw_cfg
```
Comment (andrisaar) [inline stage0/src/initramfs.rs:47]: Two thoughts here:
a) If we want to support loading the kernel via the fw_cfg interface,  then we should mark the memory as reserved in the E820 map. (Or, as I mentioned below, check what SeaBIOS etc are doing in this case.)
b) If we load the kernel via the loader device, this address won't tell us much, as the loader has already laid out the file in memory.

tl;dr: I'm not sure how much we can trust and/or use the address and size here.

File: stage0/src/initramfs.rs (hunk 1 above)
Comment (conradgrobler) [reply stage0/src/initramfs.rs:91]: They don't mark it as reserved, and I also think we should not. This is a temporary compressed archive. The kernel extracts it to somewhere else in memory before actually mounting it as a ramdisk.

**Review by andrisaar — APPROVED** [review]

File: stage0/src/initramfs.rs (hunk 2 above)
Comment (conradgrobler) [reply stage0/src/initramfs.rs:47]: If the kernel is an ELF file QEMU parses it and provides the already laid-out memory snapshot here, not the ELF file, and is loaded in the location specified by the ELF headers. Based on the testing I have done these values seem to match the result of loading it with a loader device.

I am currently considering also loading the kernel via the fw_cfg device, but even then I don't think we should mark it as reserved since we don't mark it as reserved when we load it via the loader device. Also, if we load a compressed image rather than an ELF file, it extracts the inner ELF and loads it elsewhere and then this memory can actually be reused.
